torchgeo/datasets/sentinel.py

```diff
@@ -25,6 +25,185 @@ class Sentinel(RasterDataset):
+
+    # ASF DAAC GRD RTC uses a different naming scheme (README.md.txt in download):
+    #
+    # S1x_yy_aaaaaaaaTbbbbbb_ppo_RTCzz_u_defklm_ssss
+    #
+    # x:          Sentinel-1 Mission (A or B)
+    # yy:         Beam Mode
+    # aaaaaaaa:   Start Date of Acquisition (YYYYMMDD)
+    # bbbbbb:     Start Time of Acquisition (HHMMSS)
+    # pp:         Polarization: Dual-pol (D) vs. Single-pol (S),
+    #                 primary polarization (H vs. V)
+    # o:          Orbit Type: Precise (P), Restituted (R), or Original Predicted (O)
+    # zz:         Terrain Correction Pixel Spacing
+    # u:          Software Package Used: GAMMA (G)
+    # d:          Gamma-0 (g) or Sigma-0 (s) Output
+    # e:          Power (p) or Decibel (d) or Amplitude (a) Output
+    # f:          Unmasked (u) or Water Masked (w)
+    # k:          Not Filtered (n) or Filtered (f)
+    # l:          Entire Area (e) or Clipped Area (c)
+    # m:          Dead Reckoning (d) or DEM Matching (m)
+    # ssss:       Product ID
+    filename_glob = "S1*{}.*"
+    filename_regex = r"""
+        ^S1(?P<mission>[AB])
+        _(?P<mode>SM|IW|EW|WV)
+        _(?P<date>\d{8}T\d{6})
+        _(?P<polarization>[DS][HV])
+        (?P<orbit>[PRO])
+        _RTC(?P<spacing>\d{2})
+        _(?P<package>G)
+        _(?P<output1>[gs])
+        (?P<output2>[pda])
+        (?P<mask>[uw])
+        (?P<filter>[nf])
+        (?P<area>[ec])
+        (?P<matching>[dm])
+        _(?P<product>[0-9A-Z]{4})
+        _(?P<band>[VH]{2})
+        \.
+    """
+    date_format = "%Y%m%dT%H%M%S"
+    all_bands = ["HH", "HV", "VV", "VH"]
+    separate_files = True
+
+    def __init__(
+        self,
+        root: str = "data",
+        crs: Optional[CRS] = None,
+        res: float = 10,
+        bands: Sequence[str] = ["VV", "VH"],
+        transforms: Optional[Callable[[Dict[str, Any]], Dict[str, Any]]] = None,
+        cache: bool = True,
+    ) -> None:
+        """Initialize a new Dataset instance.
+
+        Args:
+            root: root directory where dataset can be found
+            crs: :term:`coordinate reference system (CRS)` to warp to
+                (defaults to the CRS of the first file found)
+            res: resolution of the dataset in units of CRS
+                (defaults to the resolution of the first file found)
+            bands: bands to return (defaults to all bands)
+            transforms: a function/transform that takes an input sample
+                and returns a transformed version
+            cache: if True, cache file handle to speed up repeated sampling
+
+        Raises:
+            AssertionError: if ``bands`` is invalid
+            FileNotFoundError: if no files are found in ``root``
+        """
+        # Sentinel-1 uses dual polarization, it can only transmit either
+        # horizontal or vertical at a single time
+        assert list(bands) == ["HH", "HV"] or list(bands) == ["VV", "VH"]
```

Comment:
This is overly restrictive, but it's done for two reasons:

1. `RasterDataset` requires `bands` to be present for all scenes, you can't have some scenes with HH/HV and other scenes with VV/VH. The easiest way to support that would be:
    ```python
    Sentinel1(root, bands=["HH", "HV"]) | Sentinel1(root, bands=["VV", "VH"])
    ```
2. It greatly simplifies plotting if we assume we always have both bands and they are in a specific order

Will anyone be upset if they can't use only a single band or can't specify them in reverse order?

Reply:
Bands shouldn't have to be in a specific order as you are given the order in `self.bands`

Reply:
True, but it adds a bit of extra code:
```python
if self.bands[0] == "HH" or self.bands[0] == "VV":
    hh = self.bands[0]
    hv = self.bands[1]
else:
    hh = self.bands[1]
    hv = self.bands[0]
```
and it isn't clear to me if it's worth it.

Reply:
These 6 lines aren't some complicated hack / keeping them out isn't a great simplification. I would keep plotting working regardless of band order or make some enum that specifies which subset of options we accept for bands

Reply:
Okay, check the latest commit. I relaxed pretty much all constraints so that the following are supported:

* single band: [HH], [HV], [VV], [VH]
* dual band: [HH, HV], [VV, VH]
* dual band (reverse order): [HV, HH], [VH, VV]

but the following are still invalid:

* no bands: [] - doesn't make sense, impossible to support
* duplicate bands: [HH, HH] - doesn't make sense, but possible to support
* quad bands: [HH, HV, VV, VH] - not possible to support with current RasterDataset implementation

These are supported both for data loading and plotting. I also added a ton of new tests to check for these.

Only remaining thing we could potentially add support for is a plotting option to choose between RGB or dual grayscale for dual band plotting. But I didn't want to make things any more complicated than they are, and we don't support this for any other dataset at the moment.

Reply:
THis seems completely reasonable!